File: python/src/actr_harness/services/neuro_core/conditions/symbolic_matcher.py
from typing import Any

from ..buffers_view import BuffersView


class SymbolicMatcher:
    def evaluate(self, condition_node: dict[str, Any], view: BuffersView) -> bool:
        t = condition_node.get("type")
        if t == "and":
            return all(self.evaluate(c, view) for c in condition_node["conditions"])
        if t == "or":
            return any(self.evaluate(c, view) for c in condition_node["conditions"])
        if t == "not":
            return not self.evaluate(condition_node["condition"], view)
        if t == "equals":
            return view.get(condition_node["slot"]) == condition_node["value"]
        if t == "exist":
            return view.get(condition_node["slot"]) is not None
        return False

File: Dotnet/src/Harness.Core/NeuroCore/SymbolicMatcher.cs
namespace Harness.Core.NeuroCore;

public class SymbolicMatcher
{
    public bool Evaluate(IReadOnlyDictionary<string, object?> conditionNode, BuffersView view)
    {
        if (!conditionNode.TryGetValue("type", out var typeValue) || typeValue is not string type)
        {
            return false;
        }

        return type switch
        {
            "and" => EnumerateConditions(conditionNode, "conditions").All(condition => Evaluate(condition, view)),
            "or" => EnumerateConditions(conditionNode, "conditions").Any(condition => Evaluate(condition, view)),
            "not" => conditionNode.TryGetValue("condition", out var nested)
                     && nested is IReadOnlyDictionary<string, object?> nestedMap
                     && !Evaluate(nestedMap, view),
            "equals" => conditionNode.TryGetValue("slot", out var slotValue)
                        && slotValue is string slot
                        && conditionNode.TryGetValue("value", out var expectedValue)
                        && Equals(view.Get(slot), expectedValue),
            "exist" => conditionNode.TryGetValue("slot", out var existsSlotValue)
                       && existsSlotValue is string existsSlot
                       && view.Get(existsSlot) is not null,
            _ => false
        };
    }

    private static IEnumerable<IReadOnlyDictionary<string, object?>> EnumerateConditions(
        IReadOnlyDictionary<string, object?> node,
        string key)
    {
        if (!node.TryGetValue(key, out var value) || value is not IEnumerable<object?> items)
        {
            return [];
        }

        return items.OfType<IReadOnlyDictionary<string, object?>>();
    }
}
